### PR TITLE
More efficient identity and multiplier weight to fee

### PR DESCRIPTION
**Pull request by nazar-pc**
#10785 left me with impression that `ConstantMultiplier` implementation is too heavy for what is needed in that particular case, so I decided to apply small optimization to it and to `IdentityFee` to avoid going polynomial route since it is not necessary for those cases.

### CODE REVIEW

File: frame/support/src/weights.rs [hunk 1]
```diff
@@ -710,6 +710,10 @@ where
 			degree: 1,
 		})
 	}
+
+	fn calc(weight: &Weight) -> Self::Balance {
```
Comment (kianenigma) [inline frame/support/src/weights.rs:714]: I find this API a bit strange, that now `polynomial` still needs to be implemented but it is not really used, right? Nonetheless, looks like a reasonable micro optimization.

**Review by kianenigma — APPROVED** [review]

**Comment by shawntabrizi** [comment]
Sorry, I don't get it. Where is `calc` being used in this PR?

Is this actually a measurable optimization?

Seems like API confusion is not worth micro-optimizations.

**Comment by nazar-pc** [comment]
> Sorry, I don't get it. Where is calc being used in this PR?

`calc()` is a method of `WeightToFeePolynomial` trait. Method is called for every transaction in order to convert weight to fee. without these optimizations this would have been used instead:
https://github.com/paritytech/substrate/blob/b13c73f16c1e1f7a63c83c6f8b9ce93cb05a4f92/frame/support/src/weights.rs#L668-L693

I didn't benchmark it, but I also don't see the point of doing any of that if we can just do nothing or one multiplication. API is awkward, I agree, but I didn't design it :shrug:

**Comment by shawntabrizi** [comment]
Sorry @nazar-pc I deleted my comment cause I didnt read any code, but now I did.

It is strange to me that there is a `WeightToFeePolynomial` trait at all.

Seems the trait should simply be `WeightToFee` with a single function `fn calc` that should be implemented, and this is what should be required in the pallet.

Then there should be a subtrait `WeightToFeePolynomial`, or just a struct, which implements `WeightToFee` but in the case of a polynomial.

Does that make sense? Then you can simply have Identity be the true identity without any polynomial bs.

**Comment by nazar-pc** [comment]
I like the idea of `WeightToFee`. I can make an alternative PR when I have time to do so if that'd be preferred.

**Comment by shawntabrizi** [comment]
I think you just need this:

```
/// A trait that describes a general weight to fee calculation.
pub trait WeightToFee {
	/// The type that is returned as result from polynomial evaluation.
	type Balance: BaseArithmetic + From<u32> + Copy + Unsigned;

	fn calc(weight: &Weight) -> Self::Balance;
}

pub struct WeightToFeePolynomialWrapper<W>(sp_std::marker::PhantomData::<W>);

impl<W> WeightToFee for WeightToFeePolynomialWrapper<W>
	where
		W: WeightToFeePolynomial,
{
	type Balance = W::Balance;

	fn calc(weight: &Weight) -> Self::Balance {
		W::calc(weight)
	}
}
```

Then just update the `TransactionPayment` Pallet

**Comment by shawntabrizi** [comment]
Seems in TransactionPayment Pallet, there is some exposed consts which assume that we use `polynomial`. @kianenigma is this needed?

**Review by shawntabrizi — APPROVED** [review]
anyway, my comments aren't blocking this optimization. so lets get this in

**Comment by shawntabrizi** [comment]
bot merge

**Comment by shawntabrizi** [comment]
@nazar-pc this one is a bit too baby to get a tip, but happy to send a tip if you do the changes mentioned in the conversation above :)

**Comment by kianenigma** [comment]
> Seems in TransactionPayment Pallet, there is some exposed consts which assume that we use `polynomial`. @kianenigma is this needed?

They are only for UI, and I agree that we should break them and not assume the existence of a polynomial anymore. 

All UIs should use the payment RPC for fee estimation.

**Pull request merged**